File: src/leanctx_sdk/receipt.py
"""Truthful host/evaluator receipt projection."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Mapping, Optional, Tuple

from .errors import ArtifactIntegrityError, ValidationError
from .protocol import (
    ContextSource,
    ContextView,
    HostOutcome,
    Integrity,
    SCHEMA_VERSION,
    _freeze,
    _plain,
    _text,
    canonical_bytes,
    validate_ref,
)


def _safe_usage(value: Optional[Mapping[str, object]]) -> Optional[Mapping[str, object]]:
    if value is None:
        return None
    if not isinstance(value, Mapping):
        raise ValidationError("usage must be a mapping")
    frozen = _freeze(dict(value))
    # Force a deterministic, finite JSON representation and prevent arbitrary
    # host objects from crossing into the receipt projection.
    try:
        canonical_bytes(_plain(frozen))
    except ValidationError as exc:
        raise ValidationError("usage must be deterministic JSON data") from exc
    return frozen


@dataclass(frozen=True)
class ContextReceipt:
    session_id: str
    task_id: str
    plan_id: Optional[str]
    view: Optional[ContextView]
    outcome: str
    integrity_status: str
    degradations: Tuple[str, ...] = ()
    usage: Optional[Mapping[str, object]] = None
    host_exception_type: Optional[str] = None
    host_result: object = field(default=None, repr=False, compare=False)
    host_exception: Optional[BaseException] = field(default=None, repr=False, compare=False)
    schema_version: int = SCHEMA_VERSION

    def __post_init__(self):
        _text(self.session_id, "session_id", 512)
        _text(self.task_id, "task_id", 512)
        if self.plan_id is not None:
            validate_ref(self.plan_id, "plan_id")
            if not self.plan_id.startswith("plan:sha256:"):
                raise ValidationError("plan_id must be a deterministic plan reference")
        if self.view is not None and not isinstance(self.view, ContextView):
            raise ValidationError("receipt view must be ContextView")
        if self.outcome not in {item.value for item in HostOutcome}:
            raise ValidationError("invalid host outcome")
        if self.integrity_status not in {item.value for item in Integrity}:
            raise ValidationError("invalid integrity status")
        if self.outcome == HostOutcome.ABORTED.value and self.host_exception_type is not None:
            _text(self.host_exception_type, "host_exception_type", 512)
            if ":" in self.host_exception_type or "\n" in self.host_exception_type:
                raise ValidationError("host_exception_type must be a safe type name")
        if self.host_exception_type is not None:
            _text(self.host_exception_type, "host_exception_type", 512)
        if self.host_exception is not None:
            if not isinstance(self.host_exception, BaseException):
                raise ValidationError("host_exception must be a BaseException")
            if self.outcome != HostOutcome.ABORTED.value:
                raise ValidationError("host_exception requires an aborted outcome")
            expected_type = (
                f"{type(self.host_exception).__module__}."
                f"{type(self.host_exception).__qualname__}"
            )
            if self.host_exception_type != expected_type:
                raise ValidationError("host_exception_type does not match host_exception")
        if (
            not isinstance(self.schema_version, int)
            or isinstance(self.schema_version, bool)
            or self.schema_version != SCHEMA_VERSION
        ):
            raise ValidationError("receipt schema_version must be 1")
        degradations = tuple(self.degradations)
        if any(not isinstance(item, str) or not item for item in degradations):
            raise ValidationError("degradations must be non-empty strings")
        object.__setattr__(self, "degradations", degradations)
        object.__setattr__(self, "usage", _safe_usage(self.usage))
        if self.integrity_status == Integrity.SEALED.value:
            if self.view is None or not self.view.verify():
                raise ValidationError("sealed receipt requires verified Engine evidence")

    @property
    def sealed(self) -> bool:
        return self.integrity_status == Integrity.SEALED.value

    @property
    def status(self) -> Optional[str]:
        return self.view.status if self.view is not None else None

    @property
    def source(self) -> Optional[ContextSource]:
        return self.view.source if self.view is not None else None

    @property
    def invocation(self):
        return self.view.invocation if self.view is not None else None

    @property
    def observation(self):
        return self.view.observation if self.view is not None else None

    @property
    def receipt_link(self):
        return self.view.receipt_link if self.view is not None else None

    @property
    def recovery_ref(self) -> Optional[str]:
        return self.view.recovery_ref if self.view is not None else None

    @property
    def output_digest(self) -> Optional[str]:
        return self.view.output_digest if self.view is not None else None

    @property
    def exception(self) -> Optional[BaseException]:
        """Original host exception; excluded from serialized receipt output."""

        return self.host_exception

    def verify(self) -> bool:
        return self.sealed and self.view is not None and self.view.verify()

    def require_verified(self) -> None:
        """Fail closed when receipt evidence is absent, degraded, or unsealed."""

        if not self.verify():
            raise ArtifactIntegrityError("receipt evidence is not sealed")

    def to_dict(self) -> Mapping[str, object]:
        result = {
            "schema_version": self.schema_version,
            "session_id": self.session_id,
            "task_id": self.task_id,
            "plan_id": self.plan_id,
            "outcome": self.outcome,
            "integrity_status": self.integrity_status,
            "degradations": list(self.degradations),
            "usage": _plain(self.usage) if self.usage is not None else None,
            "host_exception_type": self.host_exception_type,
            "status": self.status,
            "source": _plain(self.source.to_dict()) if self.source else None,
            "invocation": _plain(self.invocation) if self.invocation else None,
            "observation": _plain(self.observation) if self.observation else None,
            "receipt_link": _plain(self.receipt_link.to_dict()) if self.receipt_link else None,
            "recovery_ref": self.recovery_ref,
            "output_digest": self.output_digest,
        }
        return result


__all__ = ["ContextReceipt"]

File: packages/rust/src/receipt.rs
use serde_json::{Map, Value};

use crate::errors::{boxed, ArtifactIntegrityError, SdkResult, ValidationError};
use crate::protocol::{
    canonical_bytes, validate_ref, validate_text, ContextSource, ContextView, HostOutcome,
    Integrity, MAX_REF_BYTES, SCHEMA_VERSION,
};

#[derive(Clone, Debug, PartialEq)]
pub struct ContextReceipt {
    session_id: String,
    task_id: String,
    plan_id: Option<String>,
    view: Option<ContextView>,
    outcome: HostOutcome,
    integrity_status: Integrity,
    degradations: Vec<String>,
    usage: Option<Value>,
    host_exception_type: Option<String>,
    host_result: Option<Value>,
}

impl ContextReceipt {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        session_id: impl AsRef<str>,
        task_id: impl AsRef<str>,
        plan_id: Option<String>,
        view: Option<ContextView>,
        outcome: HostOutcome,
        integrity_status: Integrity,
        degradations: Vec<String>,
        usage: Option<Value>,
        host_exception_type: Option<String>,
        host_result: Option<Value>,
    ) -> SdkResult<Self> {
        let session_id = validate_text(session_id.as_ref(), "session_id", MAX_REF_BYTES, true)?;
        let task_id = validate_text(task_id.as_ref(), "task_id", MAX_REF_BYTES, true)?;
        let plan_id = plan_id
            .map(|value| validate_ref(&value, "plan_id"))
            .transpose()?;
        if let Some(value) = &plan_id {
            if value.len() != 76 || !value.starts_with("plan:sha256:") {
                return Err(boxed(ValidationError::new(
                    "plan_id must be a deterministic plan reference",
                )));
            }
        }
        for value in &degradations {
            if value.is_empty() || value.chars().any(|character| character < '\u{20}') {
                return Err(boxed(ValidationError::new(
                    "degradations must be non-empty strings",
                )));
            }
        }
        if let Some(value) = &usage {
            canonical_bytes(value).map_err(boxed)?;
        }
        if let Some(value) = &host_exception_type {
            validate_text(value, "host_exception_type", MAX_REF_BYTES, true)?;
            if value.contains(':') {
                return Err(boxed(ValidationError::new(
                    "host_exception_type must be a safe type name",
                )));
            }
        }
        if matches!(integrity_status, Integrity::Sealed)
            && view.as_ref().map_or(true, |value| !value.verify())
        {
            return Err(boxed(ValidationError::new(
                "sealed receipt requires verified Engine evidence",
            )));
        }
        Ok(Self {
            session_id,
            task_id,
            plan_id,
            view,
            outcome,
            integrity_status,
            degradations,
            usage,
            host_exception_type,
            host_result,
        })
    }

    pub fn session_id(&self) -> &str {
        &self.session_id
    }

    pub fn task_id(&self) -> &str {
        &self.task_id
    }

    pub fn plan_id(&self) -> Option<&str> {
        self.plan_id.as_deref()
    }

    pub fn view(&self) -> Option<&ContextView> {
        self.view.as_ref()
    }

    pub fn outcome(&self) -> HostOutcome {
        self.outcome
    }

    pub fn integrity_status(&self) -> Integrity {
        self.integrity_status
    }

    pub fn degradations(&self) -> &[String] {
        &self.degradations
    }

    pub fn usage(&self) -> Option<&Value> {
        self.usage.as_ref()
    }

    pub fn host_exception_type(&self) -> Option<&str> {
        self.host_exception_type.as_deref()
    }

    pub fn host_result(&self) -> Option<&Value> {
        self.host_result.as_ref()
    }

    pub fn sealed(&self) -> bool {
        self.integrity_status == Integrity::Sealed
    }

    pub fn status(&self) -> Option<crate::protocol::EngineStatus> {
        self.view.as_ref().map(ContextView::status)
    }

    pub fn source(&self) -> Option<&ContextSource> {
        self.view.as_ref().map(ContextView::source)
    }

    pub fn invocation(&self) -> Option<&Value> {
        self.view.as_ref().map(ContextView::invocation)
    }

    pub fn observation(&self) -> Option<&Value> {
        self.view.as_ref().map(ContextView::observation)
    }

    pub fn receipt_link(&self) -> Option<&crate::protocol::ContextReceiptLink> {
        self.view.as_ref().and_then(ContextView::receipt_link)
    }

    pub fn recovery_ref(&self) -> Option<&str> {
        self.view.as_ref().and_then(ContextView::recovery_ref)
    }

    pub fn output_digest(&self) -> Option<&str> {
        self.view.as_ref().and_then(ContextView::output_digest)
    }

    pub fn verify(&self) -> bool {
        self.sealed() && self.view.as_ref().is_some_and(ContextView::verify)
    }

    pub fn require_verified(&self) -> SdkResult<()> {
        if self.verify() {
            Ok(())
        } else {
            Err(boxed(ArtifactIntegrityError::new(
                "receipt evidence is not sealed",
            )))
        }
    }

    pub fn to_dict(&self) -> Value {
        let mut object = Map::new();
        object.insert("schema_version".to_owned(), Value::from(SCHEMA_VERSION));
        object.insert(
            "session_id".to_owned(),
            Value::String(self.session_id.clone()),
        );
        object.insert("task_id".to_owned(), Value::String(self.task_id.clone()));
        object.insert(
            "plan_id".to_owned(),
            self.plan_id.clone().map_or(Value::Null, Value::String),
        );
        object.insert(
            "outcome".to_owned(),
            Value::String(self.outcome.to_string()),
        );
        object.insert(
            "integrity_status".to_owned(),
            Value::String(self.integrity_status.to_string()),
        );
        object.insert(
            "degradations".to_owned(),
            Value::Array(
                self.degradations
                    .iter()
                    .cloned()
                    .map(Value::String)
                    .collect(),
            ),
        );
        object.insert(
            "usage".to_owned(),
            self.usage.clone().unwrap_or(Value::Null),
        );
        object.insert(
            "host_exception_type".to_owned(),
            self.host_exception_type
                .clone()
                .map_or(Value::Null, Value::String),
        );
        object.insert(
            "status".to_owned(),
            self.status()
                .map_or(Value::Null, |value| Value::String(value.to_string())),
        );
        object.insert(
            "source".to_owned(),
            self.source()
                .and_then(|value| value.to_dict().ok())
                .unwrap_or(Value::Null),
        );
        object.insert(
            "invocation".to_owned(),
            self.invocation().cloned().unwrap_or(Value::Null),
        );
        object.insert(
            "observation".to_owned(),
            self.observation().cloned().unwrap_or(Value::Null),
        );
        object.insert(
            "receipt_link".to_owned(),
            self.receipt_link()
                .map(crate::protocol::ContextReceiptLink::to_value)
                .unwrap_or(Value::Null),
        );
        object.insert(
            "recovery_ref".to_owned(),
            self.recovery_ref()
                .map_or(Value::Null, |value| Value::String(value.to_owned())),
        );
        object.insert(
            "output_digest".to_owned(),
            self.output_digest()
                .map_or(Value::Null, |value| Value::String(value.to_owned())),
        );
        Value::Object(object)
    }
}
